zcash_client_sqlite/src/wallet/transparent.rs

```diff
@@ -1023,39 +1098,90 @@ pub(crate) fn transaction_data_requests<P: consensus::Parameters>(
-    let chain_tip_height = super::chain_tip_height(conn)?;
+    let chain_tip_height =
+        super::chain_tip_height(conn)?.ok_or(SqliteClientError::ChainHeightUnknown)?;
 
     // We cannot construct address-based transaction data requests for the case where we cannot
     // determine the height at which to begin, so we require that either the target height or mined
     // height be set.
-    let mut address_request_stmt = conn.prepare_cached(
-        "SELECT ssq.address, IFNULL(t.target_height, t.mined_height)
+    let mut spend_requests_stmt = conn.prepare_cached(
+        "SELECT
+            ssq.address,
+            IFNULL(t.target_height, t.mined_height)
          FROM transparent_spend_search_queue ssq
          JOIN transactions t ON t.id_tx = ssq.transaction_id
          WHERE t.target_height IS NOT NULL
          OR t.mined_height IS NOT NULL",
     )?;
 
-    let result = address_request_stmt
-        .query_and_then([], |row| {
+    let spend_search_rows = spend_requests_stmt.query_and_then([], |row| {
+        let address = TransparentAddress::decode(params, &row.get::<_, String>(0)?)?;
+        let block_range_start = BlockHeight::from(row.get::<_, u32>(1)?);
+        let max_end_height = block_range_start + DEFAULT_TX_EXPIRY_DELTA + 1;
+        Ok::<TransactionDataRequest, SqliteClientError>(
+            TransactionDataRequest::TransactionsInvolvingAddress {
+                address,
+                block_range_start,
+                block_range_end: Some(std::cmp::min(chain_tip_height + 1, max_end_height)),
+                request_at: None,
+                tx_status_filter: TransactionStatusFilter::Mined,
+                output_status_filter: OutputStatusFilter::All,
+            },
+        )
+    })?;
+
+    // Since we don't want to interpret funds that are temporarily held by an ephemeral address in
+    // the course of creating ZIP 320 transaction pair as belonging to the wallet, we will perform
+    // ephemeral address checks only for addresses that do not have an unexpired transaction
+    // associated with them in the database. If, for some reason, the second transaction in a ZIP
+    // 320 pair fails to be mined after the first transaction in the pair succeeded, we will begin
+    // including the associated ephemeral address in the set to be checked for funds only after
+    // the transaction that spends from it has expired.
+    let mut ephemeral_check_stmt = conn.prepare_cached(
+        "SELECT
+            cached_transparent_receiver_address,
+            transparent_receiver_next_check_time
+         FROM addresses
+         WHERE key_scope = :ephemeral_key_scope
+         AND NOT EXISTS (
+            SELECT 'x'
+            FROM transparent_received_outputs tro
+            JOIN transactions t ON t.id_tx = tro.transaction_id
+            WHERE tro.address_id = addresses.id
+            AND t.expiry_height > :chain_tip_height
```

Comment:
If `t` expires *at* the chain tip height then it has expired, so this test correctly excludes it.